### PR TITLE
DAPI-691 Add a blacklist option to the CE job queue daemon command

### PR DESCRIPTION
**Pull request by anaisbaunelemaire**
Pull request of the DAPI 691 - Add to the pim documentation how to blacklist some jobs in the execution queue => https://github.com/akeneo/pim-docs/pull/1345

**Description (for Contributor and Core Developer)**
The goal is to filter the execution of the jobs in the queue, adding a blacklist of job codes that we don't want to execute.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php [hunk 1]
```diff
@@ -13,17 +13,20 @@ class MassEditOperation implements BatchableOperationInterface
 {
     /** @var string The background job code to launch */
     protected $jobInstanceCode;
+    protected $whitelistedJobInstanceCode;
```
Comment (AymericPlanche) [inline src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php]: don't understand why you've modify anything in this file

File: src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/OperationJobLauncher.php [hunk 2]
```diff
@@ -29,9 +29,9 @@ class OperationJobLauncher
     protected $tokenStorage;
 
     /**
-     * @param JobLauncherInterface                  $jobLauncher
+     * @param JobLauncherInterface $jobLauncher
```
Comment (AymericPlanche) [inline src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/OperationJobLauncher.php]: same, do not add to your git staging everything, you've to select precisely what you want to commit

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Command/JobQueueConsumerCommand.php [hunk 3]
```diff
@@ -50,7 +50,8 @@ protected function configure()
             ->setName(self::COMMAND_NAME)
             ->setDescription('Launch a daemon that will consume job execution messages and launch the associated job execution in backgrounds')
             ->addOption('run-once', null, InputOption::VALUE_NONE, 'Launch only one job execution and stop the daemon once the job execution is finished')
-            ->addOption('job', 'j', InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY, 'Job instance codes that should be consumed')
+            ->addOption('whitelisted-job', 'w', InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY, 'Job instance codes that should be consumed')
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Command/JobQueueConsumerCommand.php]: we will discuss about this IRL but as it's in a 3.2 patch we don't want to change the command "interface"

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Command/JobQueueConsumerCommand.php [hunk 4]
```diff
@@ -174,3 +177,5 @@ private function getJobExecutionManager(): JobExecutionManager
         return $this->getContainer()->get('akeneo_batch_queue.manager.job_execution_manager');
     }
 }
+
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Command/JobQueueConsumerCommand.php]: do not commit things like this

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php [hunk 5]
```diff
@@ -40,15 +40,18 @@ public function publish(JobExecutionMessage $jobExecutionMessage): void
     /**
      * {@inheritdoc}
      */
-    public function consume(string $consumer, array $jobInstanceCodes = []): JobExecutionMessage
+    public function consume(string $consumer, array $jobInstanceCodes = [], array $blacklistedJobInstanceCodes = []): JobExecutionMessage
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php]:
```suggestion
    public function consume(string $consumer, array $whitelistedJobInstanceCodes = [], array $blacklistedJobInstanceCodes = []): JobExecutionMessage
```

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php [hunk 6]
```diff
@@ -40,15 +40,18 @@ public function publish(JobExecutionMessage $jobExecutionMessage): void
...
     {
         $hasBeenUpdated = false;
+        $jobExecutionMessage = '';
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php]: you change the behavior of the script, see line 57 the condition will never be `null`

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php [hunk 7]
```diff
@@ -41,6 +42,7 @@ public function __construct(
 
     /**
      * @param JobExecutionMessage $jobExecutionMessage
+     * @throws DBALException
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: you've applied a recommendation of your editor to add those lines automagically ?

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php [hunk 8]
```diff
@@ -143,8 +148,7 @@ public function getAvailableJobExecutionMessageFilteredByCodes(array $jobInstanc
     q.consumer IS NULL
 AND ji.code IN (:job_instance_codes)
 ORDER BY
-    q.create_time, id
-LIMIT 1;
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: why do you remove the limit ?

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php [hunk 9]
```diff
@@ -157,12 +161,47 @@ public function getAvailableJobExecutionMessageFilteredByCodes(array $jobInstanc
...
+    q.id, q.job_execution_id, q.create_time, q.updated_time, q.options, q.consumer
+FROM
+    akeneo_batch_job_execution_queue q
+INNER JOIN akeneo_batch_job_execution je ON je.id = q.job_execution_id
+INNER JOIN akeneo_batch_job_instance ji ON ji.id = je.job_instance_id
+WHERE
+    q.consumer IS NULL
+AND ji.code NOT IN (:blacklisted_job_instance_codes)
+ORDER BY
+    q.create_time, id;
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: where is the limit ?

File: src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php [hunk 10]
```diff
@@ -27,7 +27,8 @@ public function publish(JobExecutionMessage $jobExecutionMessage): void;
      * @param string $consumer name of the consumer
      * @param string[] $jobInstanceCodes name of the job instances
      *
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php]: cs

File: yarn.lock [hunk 11]
```diff
@@ -646,6 +646,23 @@
     exec-sh "^0.3.2"
     minimist "^1.2.0"
 
+"@emotion/is-prop-valid@^0.8.1":
```
Comment (AymericPlanche) [inline yarn.lock]: don't think you've to commit this change regarding the scope of your adaptation

File: src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php (hunk 1 above)
Comment (anaisbaunelemaire) [reply src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php]: When I coded, I tested a lot of stuff, and I forgot to remove this modification, which is useless, I grant you.

File: src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/OperationJobLauncher.php (hunk 2 above)
Comment (anaisbaunelemaire) [reply src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/OperationJobLauncher.php]: I do not understand what you mean. We will talk about it.

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Command/JobQueueConsumerCommand.php (hunk 3 above)
Comment (anaisbaunelemaire) [reply src/Akeneo/Tool/Bundle/BatchQueueBundle/Command/JobQueueConsumerCommand.php]: I understand. I didn't think about that.

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php (hunk 7 above)
Comment (anaisbaunelemaire) [reply src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: Yes I think.

File: src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php [hunk 12]
```diff
@@ -44,10 +47,16 @@ public function getBatchConfig(): array
     }
 
     /**
-     * {@inheritdoc}
+     * Get the code of the JobInstance
+     *
+     * @return string
      */
-    public function getJobInstanceCode(): string
+    public function getJobInstanceCode()
```
Comment (pierrejolly) [inline src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php]: Why do you remove the typehint ?

Comment (anaisbaunelemaire) [reply src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php]: A mistake. Corrected in the last commit.

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Command/JobQueueConsumerCommand.php [hunk 13]
```diff
@@ -128,6 +131,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
      * @param JobExecutionMessage $jobExecutionMessage
      *
      * @return array
+     * @throws \Doctrine\DBAL\DBALException
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Command/JobQueueConsumerCommand.php]: useless

File: src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/OperationJobLauncher.php (hunk 2 above)
Comment (AymericPlanche) [inline src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/OperationJobLauncher.php]: all the changes of this file is useless, pick the one of the 3.2 branch

File: src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php [hunk 14]
```diff
@@ -44,10 +47,16 @@ public function getBatchConfig(): array
...
     {
         return $this->jobInstanceCode;
     }
+
```
Comment (AymericPlanche) [inline src/Akeneo/Pim/Enrichment/Bundle/MassEditAction/Operation/MassEditOperation.php]: why all those blank lines

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php [hunk 15]
```diff
@@ -174,6 +210,7 @@ public function getJobInstanceCode(JobExecutionMessage $jobExecutionMessage): ?s
 JOIN akeneo_batch_job_instance ji ON ji.id = je.job_instance_id
 WHERE 
     je.id = :id
+    LIMIT 1;
```
Comment (AymericPlanche) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]:
```suggestion
LIMIT 1;
```

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php [hunk 16]
```diff
@@ -40,15 +40,22 @@ public function publish(JobExecutionMessage $jobExecutionMessage): void
     /**
      * {@inheritdoc}
      */
-    public function consume(string $consumer, array $jobInstanceCodes = []): JobExecutionMessage
+    public function consume(string $consumer, array $whitelistedJobInstanceCodes = [], array $blacklistedJobInstanceCodes = []): JobExecutionMessage
```
Comment (pierallard) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php]: BC! Add it to CHANGELOG please.

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php [hunk 17]
```diff
@@ -40,15 +40,22 @@ public function publish(JobExecutionMessage $jobExecutionMessage): void
...
     {
         $hasBeenUpdated = false;
+        $jobExecutionMessage = null;
+
+        if (!empty($whitelistedJobInstanceCodes) && !empty($blacklistedJobInstanceCodes)) {
+            throw new \InvalidArgumentException('You can not use a whitelist filter and a blacklist filter at the same time');
```
Comment (pierallard) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php]: "cannot"

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/spec/Queue/DatabaseJobExecutionQueueSpec.php [hunk 18]
```diff
@@ -28,29 +28,52 @@ function it_publishes_a_job_execution_message(
...
+        $jobExecutionMessageRepository->getAvailableJobExecutionMessageFilteredByCodes(['csv_export_product'])->shouldNotBeCalled();
+        $jobExecutionMessageRepository->getAvailableNotBlacklistedJobExecutionMessageFilteredByCodes(['csv_export_product'])->willReturn($jobExecutionMessage);
+        $jobExecutionMessageRepository->updateJobExecutionMessage($jobExecutionMessage)->willReturn(true);
+
+        $jobExecutionMessage->consumedBy('consumer_name')->shouldBeCalled();
+
+        $this->consume('consumer_name', [], ['csv_export_product'])->shouldReturn($jobExecutionMessage);
+    }
+
+    function it_consumes_a_job_execution_message_with_both_whitelist_and_blacklist_filters(
```
Comment (pierallard) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/spec/Queue/DatabaseJobExecutionQueueSpec.php]: it_throws_an_exception_when_trying_to...

File: src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php [hunk 19]
```diff
@@ -25,9 +25,10 @@ public function publish(JobExecutionMessage $jobExecutionMessage): void;
      * This method loops until there is a message to consume into the queue.
      *
      * @param string $consumer name of the consumer
-     * @param string[] $jobInstanceCodes name of the job instances
+     * @param string[] $whitelistedJobInstanceCodes name of the whitelisted job instances
```
Comment (pierallard) [inline src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php]: codes of the ... (same under)

File: src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php [hunk 20]
```diff
@@ -25,9 +25,10 @@ public function publish(JobExecutionMessage $jobExecutionMessage): void;
...
+     * @param string[] $blacklistedJobInstanceCodes name of the blacklisted job instances
      *
      * @return JobExecutionMessage
      */
-    public function consume(string $consumer, array $jobInstanceCodes = []): JobExecutionMessage;
+    public function consume(string $consumer, array $whitelistedJobInstanceCodes = [], array $blacklistedJobInstanceCodes = []): JobExecutionMessage;
```
Comment (pierallard) [inline src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php]: BC too :)

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php [hunk 21]
```diff
@@ -157,6 +158,41 @@ public function getAvailableJobExecutionMessageFilteredByCodes(array $jobInstanc
         return false !== $row ? $this->jobExecutionHydrator->hydrate($row) : null;
     }
 
+    /**
+     * Gets a job execution message which is not blacklisted and that has not been consumed yet
+     * If there is no job execution available, it returns null.
+     *
+     * @param array $blacklistedJobInstanceCodes
+     *
+     * @return JobExecutionMessage|null
```
Comment (BitOne) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: Not needed as it's part of the function signature now.

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php [hunk 22]
```diff
@@ -157,6 +158,41 @@ public function getAvailableJobExecutionMessageFilteredByCodes(array $jobInstanc
         return false !== $row ? $this->jobExecutionHydrator->hydrate($row) : null;
     }
 
+    /**
+     * Gets a job execution message which is not blacklisted and that has not been consumed yet
+     * If there is no job execution available, it returns null.
+     *
+     * @param array $blacklistedJobInstanceCodes
```
Comment (BitOne) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: Not needed as it's part of the function signature now.

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php [hunk 23]
```diff
@@ -157,6 +158,41 @@ public function getAvailableJobExecutionMessageFilteredByCodes(array $jobInstanc
...
+     *
+     * @return JobExecutionMessage|null
+     */
+    public function getAvailableNotBlacklistedJobExecutionMessageFilteredByCodes(array $blacklistedJobInstanceCodes): ?JobExecutionMessage
```
Comment (BitOne) [inline src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: I think we can something shorter and maybe not linked to the blacklisted codes.

Maybe even better, a specific query class?

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php (hunk 16 above)
Comment (nmarniesse) [reply src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/DatabaseJobExecutionQueue.php]: Not a BC, the parameter is just renamed if I well understand. And the third param is optionnal

File: src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php [hunk 24]
```diff
@@ -25,9 +25,10 @@ public function publish(JobExecutionMessage $jobExecutionMessage): void;
      * This method loops until there is a message to consume into the queue.
      *
      * @param string $consumer name of the consumer
-     * @param string[] $jobInstanceCodes name of the job instances
+     * @param string[] $whitelistedJobInstanceCodes name of the whitelisted job instances
+     * @param string[] $blacklistedJobInstanceCodes name of the blacklisted job instances
```
Comment (nmarniesse) [inline src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php]: IMO we should modify the comment of this method and explain how it works. If I well understand, if the blacklist jobs are taken in account only if whitelist is empty

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php (hunk 23 above)
Comment (anaisbaunelemaire) [reply src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: We suggest thinking about grouping requests in the same file for the next version. Is it good for you?
It's true that the name of this function is long ... I'm thinking about renaming it ;-) But how? I took the name of the previous function ( getAvailableJobExecutionMessageFilteredByCodes() ), which had been built before.

File: src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php (hunk 24 above)
Comment (anaisbaunelemaire) [reply src/Akeneo/Tool/Component/BatchQueue/Queue/JobExecutionQueueInterface.php]: Here is what this method does, in detail:

If no whitelist or blacklist is passed as a parameter, we get the last job execution message from the queue, which is not consumed yet.

If a whitelist is passed as a parameter, we get the last job execution message from the queue, which is not consumed yet and which is in this whitelist.

If a blacklist is passed as a parameter, we get the last job execution message from the queue, which is not consumed yet and which is not in this blacklist.

This method loops until there is a message to consume in the queue.

File: src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php (hunk 23 above)
Comment (BitOne) [reply src/Akeneo/Tool/Bundle/BatchQueueBundle/Queue/JobExecutionMessageRepository.php]: Yes, grouping the queries in a different file would be nice.

Okay for the naming if it's coherent with the existing code.

**Pull request merged**